pkg/controller/admissionchecks/multikueue/controller.go

```diff
@@ -0,0 +1,382 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kueuealpha "sigs.k8s.io/kueue/apis/kueue/v1alpha1"
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/util/admissioncheck"
+)
+
+const (
+	ControllerName = "kueue.x-k8s.io/multikueue"
```

Comment:
Do we have a use case for supporting multiple AdmissionChecks with this controller? If now, I would suggest to verify the name of the AdmissionCheck to be MultiKueue, as in the KEP.

Reply:
IIUC the KEP says about one named MultiKueue [here](https://github.com/kubernetes-sigs/kueue/tree/main/keps/693-multikueue#summary). We are not really testing multiple admission checks in e2e or integration tests currently. It is always easier to relax validation that strenghten, so if not needed currently, I would suggest validating the name in webhook.

Reply:
Multikueue is a type of admission check , just like the provisioning one, in practice we can have multiple admission checks defined in a cluster each o them added to a specific set of cluster queues.

Reply:
As discussed online, we will support multiple admission checks with the same controllerName, but different names, so that different Admisssion Checks can be used for different clusterqueues, with potentially different subset of clusters. It is not clear if we need multiple admission checks for one clusterqueue. However, we cannot validate globally. Feel free to resolve.